crates/uv-distribution-filename/src/expanded_tags.rs

```diff
@@ -0,0 +1,459 @@
+    #[error("The wheel tag \"{0}\" contains too many segments")]
+    ExtraSegment(String),
+    #[error("The wheel tag \"{0}\" contains an invalid language tag")]
+    InvalidLanguageTag(String, #[source] ParseLanguageTagError),
+    #[error("The wheel tag \"{0}\" contains an invalid ABI tag")]
+    InvalidAbiTag(String, #[source] ParseAbiTagError),
+    #[error("The wheel tag \"{0}\" contains an invalid platform tag")]
+    InvalidPlatformTag(String, #[source] ParsePlatformTagError),
+}
+
+/// Parse an expanded (i.e., simplified) wheel tag, e.g. `py3-none-any`.
+///
+/// Unlike parsing tags in a wheel filename, each tag in this case is expected to contain exactly
+/// three segments separated by `-`: a language tag, an ABI tag, and a platform tag; however,
+/// empirically, some build backends do emit multipart tags (like `cp312-cp312-manylinux_2_17_x86_64.manylinux2014_x86_64`),
+/// so we allow those too.
+fn parse_expanded_tag(tag: &str) -> Result<WheelTag, ExpandedTagError> {
+    let mut splitter = memchr::Memchr::new(b'-', tag.as_bytes());
+    if tag.is_empty() {
+        return Err(ExpandedTagError::MissingLanguageTag(tag.to_string()));
+    }
+    let Some(python_tag_index) = splitter.next() else {
+        return Err(ExpandedTagError::MissingAbiTag(tag.to_string()));
+    };
+    let Some(abi_tag_index) = splitter.next() else {
+        return Err(ExpandedTagError::MissingPlatformTag(tag.to_string()));
+    };
+    if splitter.next().is_some() {
+        return Err(ExpandedTagError::ExtraSegment(tag.to_string()));
+    }
+
+    let python_tag = &tag[..python_tag_index];
+    let abi_tag = &tag[python_tag_index + 1..abi_tag_index];
+    let platform_tag = &tag[abi_tag_index + 1..];
+
+    let is_small = memchr(b'.', tag.as_bytes()).is_none();
+
+    if let Some(small) = is_small
+        .then(|| {
+            Some(WheelTagSmall {
+                python_tag: LanguageTag::from_str(python_tag).ok()?,
+                abi_tag: AbiTag::from_str(abi_tag).ok()?,
+                platform_tag: PlatformTag::from_str(platform_tag).ok()?,
+            })
+        })
+        .flatten()
+    {
+        Ok(WheelTag::Small { small })
+    } else {
+        Ok(WheelTag::Large {
+            large: Box::new(WheelTagLarge {
+                build_tag: None,
+                python_tag: MemchrSplitter::split(python_tag, b'.')
+                    .map(LanguageTag::from_str)
+                    .filter_map(Result::ok)
+                    .collect(),
+                abi_tag: MemchrSplitter::split(abi_tag, b'.')
+                    .map(AbiTag::from_str)
+                    .filter_map(Result::ok)
+                    .collect(),
+                platform_tag: MemchrSplitter::split(platform_tag, b'.')
+                    .map(PlatformTag::from_str)
+                    .filter_map(Result::ok)
+                    .collect(),
+                repr: tag.into(),
+            }),
+        })
+    }
+}
+
+#[cfg(test)]
+mod tests {
```

Comment:
Can you add a test for an invalid segment, that we're using `repr` to not lose it?